src/filters.rs

```diff
@@ -0,0 +1,272 @@
+    Variable(Variable),
+    Text(Text),
+    TranslatedText(Text),
+    Int(BigInt),
+    Float(f64),
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Argument {
+    pub at: (usize, usize),
+    pub argument_type: ArgumentType,
+}
+
+impl Render for Argument {
+    fn resolve<'t, 'py>(
+        &self,
+        py: Python<'py>,
+        template: TemplateString<'t>,
+        context: &mut Context,
+    ) -> TemplateResult<'t, 'py> {
+        Ok(Some(match &self.argument_type {
+            ArgumentType::Text(text) => return text.resolve(py, template, context),
+            ArgumentType::TranslatedText(_text) => todo!(),
+            ArgumentType::Variable(variable) => match variable.resolve(py, template, context)? {
+                Some(content) => content,
+                None => {
+                    let key = template.content(variable.at).to_string();
+                    let context: HashMap<&String, &Bound<'py, PyAny>> = context
+                        .context
+                        .iter()
+                        .map(|(k, v)| (k, v.bind(py)))
+                        .collect();
+                    let object = format!("{:?}", context);
+                    return Err(RenderError::VariableDoesNotExist {
+                        key,
+                        object,
+                        key_at: variable.at.into(),
+                        object_at: None,
+                    }
+                    .into());
+                }
+            },
+            ArgumentType::Float(number) => Content::Float(*number),
+            ArgumentType::Int(number) => Content::Int(number.clone()),
+        }))
+    }
+}
+
+#[derive(Debug)]
+pub enum FilterType {
+    Add(Argument, AddFilter),
+    AddSlashes(AddSlashesFilter),
+    Capfirst(CapfirstFilter),
+    Default(Argument, DefaultFilter),
+    External(Py<PyAny>, Option<Argument>, ExternalFilter),
+    Lower(LowerFilter),
+}
+
+pub trait Applicable<'t, 'py> {
+    fn apply(variable: Option<Content<'t, 'py>>, context: &mut Context) -> TemplateResult<'t, 'py>;
+}
+
+pub trait ApplicableArg<'t, 'py> {
+    fn apply(
+        variable: Option<Content<'t, 'py>>,
+        arg: &Argument,
+        py: Python<'py>,
+        template: TemplateString<'t>,
+        context: &mut Context,
+    ) -> TemplateResult<'t, 'py>;
+}
+
+pub trait ApplicableFilter<'t, 'py> {
+    fn apply(
+        filter: &Py<PyAny>,
+        variable: Option<Content<'t, 'py>>,
+        arg: Option<&Argument>,
+        py: Python<'py>,
+        template: TemplateString<'t>,
+        context: &mut Context,
+    ) -> TemplateResult<'t, 'py> {
+        let arg = match arg {
+            Some(arg) => arg.resolve(py, template, context)?,
+            None => None,
+        };
+        let filter = filter.bind(py);
+        let value = match arg {
+            Some(arg) => filter.call1((variable, arg))?,
+            None => filter.call1((variable,))?,
+        };
+        Ok(Some(Content::Py(value)))
+    }
+}
```

Comment:
I think we can probably simplify this down to one new trait:

```rust
pub trait ResolveFilter {
    fn resolve<'t, 'py>(
        &self,
        py: Python<'py>,
        template: TemplateString<'t>,
        context: &mut Context,
        left: Option<Content<'t, 'py>>,
    ) -> TemplateResult<'t, 'py>;
```

To do this, we'd need to move `Argument` into the `AddFilter` and `DefaultFilter` structs. And similarly move `Py<PyAny>` and `Option<Argument>` into `ExternalFilter`.

Reply:
I agree this is a lot nicer.